metagraph/src/annotation/taxonomy/tax_classifier.hpp

```diff
@@ -0,0 +1,200 @@
+     * @param [input] tax_tree_filepath ->  path to a taxonomic tree ("nodes.dmp" file).
+     * @param [optional input] label_taxid_map_filepath ->  path to the acc_version to taxid lookup table (".accession2taxid").
+     *                                                      Mandatory if the taxid is not mentioned in the label string.
+     */
+    TaxonomyClsAnno(const graph::AnnotatedDBG &anno,
+                    const double lca_coverage_rate,
+                    const double kmers_discovery_rate,
+                    const std::string &tax_tree_filepath,
+                    const std::string &label_taxid_map_filepath = "");
+    TaxonomyClsAnno() {};
+    virtual ~TaxonomyClsAnno() {};
+
+    // todo implement
+    void export_taxdb(const std::string &filepath) const;
+
+    /** assign_class_toplabels is a faster and less precise taxonomic classification approach.
+     * It takes into account only the labels with the highest frequency among the kmers in one read.
+     *
+     * Given a set of kmers taken from one read, a query on the annotation matrix will return all the labels that
+     * correspond to at least $label_fraction$ percent of the kmers.
+     * Then, returns the LCA of the taxid nodes where these frequent labels are mapping to.
+     *
+     * @param [input] sequence -> the given sequence to classify.
+     * @param [input] label_fraction -> threshold used for taxonomic classification.
+     * @return -> the classification result: a taxid node
+     */
+    TaxId assign_class_toplabels(const std::string &sequence, const double label_fraction) const;
+
+  private:
+    /**
+     * Reads and returns the taxonomic tree as a list of children.
+     *
+     * @param [input] tax_tree_filepath -> path to a "nodes.dmp" file.
+     * @param [output] tree -> tree stored as a list of children.
+     */
+    void read_tree(const std::string &tax_tree_filepath,
+                   ChildrenList *tree);
+
+    /**
+     * rmq_preprocessing computes 'this->rmq_data' field.
+     *
+     * @param [input] tree_linearization -> the linearization of the taxonomic tree.
+     */
+    void rmq_preprocessing(const std::vector<TaxId> &tree_linearization);
+
+    /**
+     * dfs_statistics method calculates the following fields:
+     *      + tree_linearization;
+     *      + this->node_depth;
+     *      + this->node_to_linearization_idx.
+     *
+     * @param [input] node -> the node that is currently processed.
+     * @param [input] tree -> the taxonomic tree stored as a list of children.
+     * @param [output] tree_linearization -> the linearization of the received tree.
+     */
+    void dfs_statistics(const TaxId node,
+                        const ChildrenList &tree,
+                        std::vector<TaxId> *tree_linearization);
+
+    TaxId find_lca(const std::vector<TaxId> &taxids) const;
+
+    std::vector<TaxId> get_lca_taxids_for_seq(const std::string_view &sequence, bool reversed) const;
```

Comment:
check which of these methods can be made `static`, and also consider moving them to the .cpp file as ordinary functions.

The same for other classes.

Reply:
I tried to create an object field only for those variables that are used in both constructor and `assign_class`. All the methods are reading or writing at least some of the object fields, thus, if we want to remove them from the object and make static/ordinary functions we will need to send the fields that are used as arguments.

e.g, for `get_accession_version_from_label` and `get_taxid_from_label` we need to send `label_type_` and `accversion_to_taxid_map_`. But `accversion_to_taxid_map_` can be empty if `LabelType=TAXID`.

Or, in case of `read_tree` we are computing both `root_node_` and `node_parent_`. I can send both of them as pointer parameters if you think that this will make the code more clear. Let me know